deployer/src/deployer/upload.py

```diff
@@ -492,6 +578,54 @@ def on_task_complete(task):
             log.info(f"Total uploaded redirects: {totals.uploaded_redirects:,} ")
         log.info(f"Total skipped files: {totals.skipped:,} matched existing S3 objects")
         log.info(f"Total upload/skip time: {upload_timer}")
+
+    if delete_unrecognized:
+        # Now `existing_bucket_objects` has mutated to only contain the keys
+        # that were not uploaded or not needed to be uploaded.
+        # That basically means all the S3 keys that exist before but are
+        # unrecognized now. For example, things that were once built but are
+        # now deleted.
+        now = datetime.datetime.utcnow().replace(tzinfo=UTC)
+        delete_keys = []
+        for key in existing_bucket_objects:
+            if "_whatsdeployed" in key:
+                # These are special and wouldn't have been uploaded
+                continue
+
+            if "/static/" in key:
+                # Careful with these!
+                # Static assets such as `main/static/js/8.0b83949c.chunk.js`
+                # are aggressively cached and they might still be referenced
+                # from within HTML pages that are still in the CDN cache.
+                # Suppose someone gets a copy of yesterday's HTML from the CDN
+                # and it refers to `/static/js/foo.abc123.js` which is not in their
+                # browser cache or the CDN's cache, what might happen is that
+                # their browser requests it even though
+                # `/static/js/foo.def456.js` is now the latest and greatest.
+                # To be safe, only delete if it's considerd "old".
+                delta = now - existing_bucket_objects[key]["LastModified"]
+                if delta.days < 30:
+                    continue
+
+            assert key.startswith(bucket_prefix)
+
+            delete_keys.append(key)
+
+        log.info(f"Total pending task deletions: {len(delete_keys):,}")
+
+        with DisplayProgress(len(delete_keys), show_progress_bar) as progress:
+
+            def on_task_complete(task):
+                progress.update(task)
+                totals.delete(task)
```

Comment:
As suggested above (i.e., assuming you go with using the `is_deletion` task attribute):
```suggestion
                totals.count(task)
```